pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -1110,6 +1116,8 @@ func newAWSCloud(config io.Reader, awsServices Services) (*Cloud, error) {
 	// Register regions, in particular for ECR credentials
 	once.Do(func() {
 		RecognizeWellKnownRegions()
+		knownRegions := KnownRegions()
+		awscredentialprovider.RegisterCredentialsProviders(knownRegions, cfg.Global.EcrRegistryId)
```

Comment:
So the logic here is that we have a hard-coded list of well-known regions, but we also discover regions e.g. from the instance metadata.  That's why we do RegisterCredentialsProviders in `RecognizeRegion`, in particular here https://github.com/kubernetes/kubernetes/blob/6d0b71740f6b7cc3c61794b01d2a8f6361e3deff/pkg/cloudprovider/providers/aws/aws.go#L1049

Reply:
Yes, that's correct. That was the existing behaviour too, this just makes a little more explicit.

The previous behaviour was that the credentialProvider was registered as a side-effect of recognising (acknowledging the existence of and building a list of) an AWS region. The previous process involved recognising the region from the metastore (in case running in custom or non-advertised region) and then the widely-known, hardcoded regions. Credential Providers were registered while that list was constructed.

This process uses `aws/regions.go` just for building up the list of regions in the same way as before (it removed the registration of the credential provider from that though). It then submits the list of all known regions and the list of accounts (if supplied in config) to `RegisterCrednetialProviders()` so that it can loop through both to create the full list of providers.